src/web/transactions.rs

```diff
@@ -23,62 +26,47 @@ pub enum TransactionDTO {
     Rejected(RejectedTransactionDTO),
 }
 
-impl TryFrom<TransactionValue> for TransactionDTO {
+impl TryFrom<TransactionQueryResult> for TransactionDTO {
     type Error = color_eyre::Report;
 
-    fn try_from(tx: TransactionValue) -> Result<Self> {
-        match tx {
-            TransactionValue::Transaction(versioned) => {
-                let tx = versioned.into_v1();
-                let base = TransactionBase::new(tx.hash(), tx.payload, tx.signatures)
-                    .wrap_err("Failed to make TransactionBase")?;
-                Ok(Self::Committed(CommittedTransactionDTO { base }))
-            }
-            TransactionValue::RejectedTransaction(versioned) => {
-                let tx = versioned.into_v1();
-                let hash = tx.hash();
-                let RejectedTransaction {
-                    payload,
-                    signatures,
-                    rejection_reason,
-                } = tx;
-                let base = TransactionBase::new(hash, payload, signatures)
-                    .wrap_err("Failed to make TransactionBase")?;
-                Ok(Self::Rejected(RejectedTransactionDTO {
-                    base,
-                    rejection_reason: rejection_reason.into(),
-                }))
-            }
+    fn try_from(tx_result: TransactionQueryResult) -> Result<Self> {
+        let TransactionValue { tx, error } = tx_result.transaction();
+        let block_hash = tx_result.block_hash();
+        let base = TransactionBase::new(tx.hash(), block_hash, tx.payload(), tx.signatures())
+            .wrap_err("Failed to make TransactionBase")?;
+
+        match error {
+            Some(rejection_reason) => Ok(Self::Rejected(RejectedTransactionDTO {
+                base,
+                rejection_reason: (*rejection_reason).clone().into(),
+            })),
+            None => Ok(Self::Committed(CommittedTransactionDTO { base })),
         }
     }
 }
 
 #[derive(Serialize)]
 struct TransactionBase {
-    hash: SerScaleHex<HashOf<SignedTransaction>>,
-    block_hash: SerScaleHex<Hash>,
-    payload: TransactionPayloadDTO,
-    signatures: BTreeSet<SignatureDTO>,
+    hash: SerScaleHex<HashOf<VersionedSignedTransaction>>,
+    block_hash: SerScaleHex<HashOf<CommittedBlock>>,
+    payload: TransactionPayload,
+    signatures: Vec<SignaturesOf<TransactionPayload>>,
 }
 
 impl TransactionBase {
-    fn new<T, U>(hash: HashOf<SignedTransaction>, payload: Payload, signatures: T) -> Result<Self>
-    where
-        T: IntoIterator<Item = U>,
-        U: Into<Signature>,
-    {
+    fn new(
+        hash: HashOf<VersionedSignedTransaction>,
+        block_hash: &HashOf<CommittedBlock>,
+        payload: &TransactionPayload,
+        signatures: &SignaturesOf<TransactionPayload>,
```

Comment:
Ah, I see what you meant by "references". I _think_ it is not a solution of the problem, because cloning is still happening, just in a different place. The point is to avoid cloning at all (so that we don't allocate extra memory, therefore we are more efficient).

I suggest you either to revert this change or to borrow it completely.

To revert means to move values to `::new()` instead of borrowing them:

```suggestion
        block_hash: HashOf<CommittedBlock>,
        payload: TransactionPayload,
        signatures: SignaturesOf<TransactionPayload>,
```

See the section below why this makes sense.

To borrow it completely means to not own the data in `TransactionBase`, but to borrow it:

```rust
struct Foo<'a> {
  bar: &'a Bar
}

impl<'a> Foo<'a> {
  fn new(bar: &'a Bar) -> Self {
    Self { bar }
  }
}

fn main() -> {
  let bar = Bar;
  // `foo` has `bar`, but doesn't own it, and there is no need for cloning.
  let foo = Foo::new(&bar);
}
```

#### Avoid extra cloning when you don't need it

Here is a more detailed example why extra cloning inside of `::new()` might be bad.

What if we don't need to clone data at all?

Consider the following code:

```rust
struct Foo {
  data: Vec<u8>
}

impl Foo {
  fn new_borrowed(data: &Vec<u8>) -> Self {
    Self { data: data.clone() }
  }

  fn new_owned(data: Vec<u8>) -> Self {
    Self { data }
  }
}

fn main() -> {
  let data: Vec<u8> = vec![1, 2, 3, 4, 5];

  // We only need to consume `data` and produce `Foo` from it.

  // Inside of `new_borrowed`, `data` will be cloned, which is unnecessary
  // It might be useful only if we want to use `data` again after this call,
  // but we can `.clone()` it here if we need
  let foo = Foo::new_borrowed(&data);

  // Here we consume data, no extra cloning
  let foo = Foo::new_owned(data)
}
```

I realise that this example is not very descriptive, and not sure if it transmits the idea I want to tell you about. Anyway, I hope so.

Reply:
@horizenight, could you comment on it?